### PR TITLE
Add module for cockroachdb

### PR DESCRIPTION
**Pull request by jsoriano**
Depends on #12465, module implemented as light module (#12270).

Collect metrics from cochroachdb, from its Prometheus endpoint.
All metrics are collected as they are and stored under `prometheus.metrics`.

An overview dashboard is added:
Captura de pantalla de 2019-06-25 21-18-14

Pending:
- [x] tests
- [ ] ~~fields.yml~~ (there will be follow up, see https://github.com/jsoriano/beats/pull/3)
- [x] dashboard

### CODE REVIEW

**Comment by exekias** [comment]
I see the pending point of `fields.yml`. This one is interesting, as already present Prometheus `fields.yml` gives us the right mapping already, still, there are no docs for the fields that we will get from the module. Is that why it's on the list?

**Comment by jsoriano** [comment]
> I see the pending point of `fields.yml`. This one is interesting, as already present Prometheus `fields.yml` gives us the right mapping already, still, there are no docs for the fields that we will get from the module. Is that why it's on the list?

Yeah, I'd like to explore the possibility of getting the documentation for these fields automatically. If not it will require at least an empty `fields.yml` to pass our tests (or we have to change the tests).

File: x-pack/metricbeat/module/cockroachdb/status/manifest.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+default: true
+input:
+  module: prometheus
```
Comment (ruflin) [inline x-pack/metricbeat/module/cockroachdb/status/manifest.yml]: LGTM for now but I would really like to see this changing to `type: prometheus` one day, meaning having real inputs.

**Comment by ruflin** [comment]
@jsoriano Is there a way to also have generated files for these light modules? This would be nice for testing, docs but also for us to see if the generated events outside the prometheus metrics contain the correct values (for example event.dataset).

File: x-pack/metricbeat/module/cockroachdb/status/manifest.yml (hunk 1 above)
Comment (jsoriano) [reply x-pack/metricbeat/module/cockroachdb/status/manifest.yml]: Yes, a possible future step after all this is to add some kind of inputs registry for implementations that don't have enough entity to be metricsets, but I didn't want to open this door yet :slightly_smiling_face:

**Comment by jsoriano** [comment]
> @jsoriano Is there a way to also have generated files for these light modules? This would be nice for testing, docs but also for us to see if the generated events outside the prometheus metrics contain the correct values (for example event.dataset).

I would use by now the tools we already have, it should be possible to use current prometheus helpers to test this module and generate the `data.json` file. I wrote a bit about testing and tooling [in the proposal](https://docs.google.com/document/d/1R_GULET8kLOozMhHdnmn68wRgKevxCfs2S2XwVMcNXw/edit#heading=h.6n5uqbwrzc4n).

File: x-pack/metricbeat/module/cockroachdb/status/manifest.yml (hunk 1 above)
Comment (ruflin) [reply x-pack/metricbeat/module/cockroachdb/status/manifest.yml]: ++

**Comment by ruflin** [comment]
@jsoriano Would be great to have the generated file directly as part of this PR to make sure we generate the right data. Great if the current tools already work.

**Comment by kaiyan-sheng** [comment]
Just curious, when there are changes made in prometheus module, for example Pablo's `le` change on histogram, will tests fail in modules like cockroachdb?

**Comment by jsoriano** [comment]
> Just curious, when there are changes made in prometheus module, for example Pablo's `le` change on histogram, will tests fail in modules like cockroachdb?

Yes, tests will fail and affected files will need to be regenerated. This would be similar to other changes we do in the framework that require regenerating golden files.

File: x-pack/metricbeat/module/cockroachdb/status/status_test.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+func init() {
+	// To be moved to some kind of helper
+	os.Setenv("BEAT_STRICT_PERMS", "false")
+	mb.Registry.SetSecondarySource(mb.NewLightModulesSource("../../../module"))
+}
+
+func TestEventMapping(t *testing.T) {
+	logp.TestingSetup()
+
+	ptest.TestMetricSet(t, "cockroachdb", "status",
```
Comment (ruflin) [inline x-pack/metricbeat/module/cockroachdb/status/status_test.go]: I wonder if this will be something we still need? Can't we just rely on the new testing framework for this? Would be great to not needed it so not golang files anymore.

Comment (jsoriano) [reply x-pack/metricbeat/module/cockroachdb/status/status_test.go]: I'd like to keep all modules testable as `go test ./module/cockroachdb/...`, what I could try to use `mbtest.TestDataFiles` instead (added in  #11818).

For the future I'd like to add the possibility of testing light modules with `metricbeat` subcommands, so anyone can develop modules just with `metricbeat`. And have a helper that does the same thing when `go test` is used for the modules in the beats repo. I mention something about this [in the light modules proposal](https://docs.google.com/document/d/1R_GULET8kLOozMhHdnmn68wRgKevxCfs2S2XwVMcNXw/edit#heading=h.6n5uqbwrzc4n).

Comment (ruflin) [reply x-pack/metricbeat/module/cockroachdb/status/status_test.go]: As long as the new testing library is used, I'm good with it. We should remove ptest soon as the new library also checks fields docs etc.

File: dev-tools/cmd/module_include_list/module_include_list.go [hunk 3]
```diff
@@ -99,19 +99,18 @@ func main() {
...
 		var foundInitMethod bool
+		goFiles, err := filepath.Glob(filepath.Join(dir, "*.go"))
+		if err != nil {
+			log.Fatalf("Failed checking for .go files in package dir: %v", err)
+		}
 		for _, f := range goFiles {
+			// Skip test files
+			if strings.HasSuffix(f, "_test.go") {
+				continue
+			}
```
Comment (jsoriano) [inline dev-tools/cmd/module_include_list/module_include_list.go]: Moved these changes to its own PR https://github.com/elastic/beats/pull/12650

**Comment by jsoriano** [comment]
Dashboard added, screenshot uploaded to issue description.

**Comment by jsoriano** [comment]
@exekias I have been playing with the possibility of generating fields.yml and docs for fields from the prometheus response, find here the changes over this PR: https://github.com/jsoriano/beats/pull/3
I think that using these fields for the mapping can be a bit excesive as they are hundreds of new fields and we already have the wildcards, but it could be nice for the documentation. But this would require more changes in the doc generator.
What do you think? Does it worth it? or I leave the fields undocumented by now?

**Comment by exekias** [comment]
Leaving them for another PR sounds reasonable to me, will comment there!

Is this one ready for review?

**Comment by jsoriano** [comment]
Opening for review, there are some things pending but they are mostly for tests and docs, we can do them in future PRs.

File: CHANGELOG.next.asciidoc [hunk 4]
```diff
@@ -235,6 +235,7 @@ https://github.com/elastic/beats/compare/v7.0.0-alpha2...master[Check the HEAD d
 - Always report Pod UID in the `pod` metricset. {pull}12345[12345]
 - Add Vsphere Virtual Machine operating system to `os` field in Vsphere virtualmachine module. {pull}12391[12391]
 - Add validation for elasticsearch and kibana modules' metricsets when xpack.enabled is set to true. {pull}12386[12386]
+- Add CockroachDB module. {pull}12465[12465]
```
Comment (kaiyan-sheng) [inline CHANGELOG.next.asciidoc]: hmmm #12467 ?

Comment (jsoriano) [reply CHANGELOG.next.asciidoc]: Correct, good catch!
